src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -505,18 +507,22 @@ static bool useHeavyMonitors() {
 // The interpreter and compiler assembly code tries to lock using the fast path
 // of this algorithm. Make sure to update that code if the following function is
 // changed. The implementation is extremely sensitive to race condition. Be careful.
+void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* locking_thread) {
```

Comment:
I suggested a new method here for the deopt thread, because I thought much of what is checked in this method is not applicable for the circumstances where we would be dealing with relocking during deopt - and thus the code would be much simpler if we had one method for the current thread, and a much simpler one for the deopt thread. Is that not the case?